pkg/config/config_test.go

```diff
@@ -0,0 +1,218 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"os"
+	"strings"
+	"syscall"
+	"testing"
+
+	"github.com/vishvananda/netlink"
+)
+
+func TestSysctlConfigEnsure(t *testing.T) {
+	mSysctl := make(map[string]string)
+
+	c := SysctlConfig{
+		Key:          "net.ipv4.conf.all.rp_filter",
+		Value:        "2",
+		DefaultValue: "1",
+		SysctlFunc: func(name string, params ...string) (string, error) {
+			mSysctl[name] = params[0]
+			return "", nil
+		},
+	}
+
+	c.Ensure(true)
+	if v := mSysctl["net.ipv4.conf.all.rp_filter"]; v == "" || v != "2" {
+		t.Error("failed to Ensure sysctlConfig rule")
+	}
+
+	c.Ensure(false)
+	if v := mSysctl["net.ipv4.conf.all.rp_filter"]; v == "" || v != "1" {
+		t.Error("failed to disable sysctlConfig rule")
+	}
+}
+
+func TestIPRouteConfigEnsure(t *testing.T) {
+	r := IPRouteConfig{
+		Route:    netlink.Route{},
+		RouteAdd: func(route *netlink.Route) error { return os.ErrExist },
+		RouteDel: func(route *netlink.Route) error { return syscall.Errno(syscall.ESRCH) },
+	}
+	if err := r.Ensure(true); err != nil {
+		t.Error("ipRouteConfig.Ensure(true) should ingore the os.ErrExist Error.")
+	}
+	if err := r.Ensure(false); err != nil {
+		t.Error("ipRouteConfig.Ensure(false) should ingore the syscall.ESRCH Error.")
+	}
+}
+
+func TestIPRuleConfigEnsure(t *testing.T) {
+	ruleList := []netlink.Rule{
+		netlink.Rule{SuppressIfgroup: -1, SuppressPrefixlen: -1, Mark: -1, Mask: -1, Goto: -1},
+		netlink.Rule{SuppressIfgroup: 1, SuppressPrefixlen: 2, Mark: -1, Mask: -1, Goto: -1},
+		netlink.Rule{SuppressIfgroup: -1, SuppressPrefixlen: -1, Mark: -1, Mask: -1, Goto: -1},
+	}
+
+	mockRuleDel := func(rule *netlink.Rule) error {
+		for i, r := range ruleList {
+			if r == *rule {
+				ruleList = append(ruleList[:i], ruleList[i+1:]...)
+				return nil
+			}
+		}
+		return nil
+	}
+
+	ipRule := IPRuleConfig{
+		Rule:     netlink.Rule{SuppressIfgroup: -1, SuppressPrefixlen: -1, Mark: -1, Mask: -1, Goto: -1},
+		RuleAdd:  func(rule *netlink.Rule) error { ruleList = append(ruleList, *rule); return nil },
+		RuleDel:  mockRuleDel,
+		RuleList: func(family int) ([]netlink.Rule, error) { return ruleList, nil },
+	}
+	if count, _ := ipRule.count(); count != 2 {
+		t.Errorf("IPRuleConfig.count() return wrong count.")
+	}
+	if count, _ := ipRule.count(); len(ruleList) != 3 || count != 2 {
+		t.Error("ruleList should contains 3 rules")
+	}
+	ipRule.Ensure(false)
+	if count, _ := ipRule.count(); len(ruleList) != 1 || count != 0 {
+		t.Error("failed to delete IPRule")
+	}
+	ipRule.Ensure(true)
+	if count, _ := ipRule.count(); len(ruleList) != 2 || count != 1 {
+		t.Error("failed to delete IPRule")
+	}
+}
+
+type MockIPtable struct {
```

Comment:
nit: s/MockIPtable/MockIPTable/

Reply:
It's more like a fake than mock. the difference is subtle tho. mock only checks if a method is called and how many times it gets called. fake mimic the behavior of the interface. and fake normally requires a test for the fake itself, before it gets plugged in test.